liberum_core/src/swarm_runner/behaviour/kademlia.rs

```diff
@@ -1,87 +1,353 @@
+    ) {
+        debug!(
+            node = self.node.name,
+            "Get providers didn't find any new records"
+        );
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(HashSet::new()).inspect_err(|e| {
+                debug!(
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+        }
+    }
+}
+
+/// Methods to handle Kademlia InboundRequest events
+impl SwarmContext {
+    fn handle_inbound_request_put_record(
+        &mut self,
+        _source: PeerId,
+        _connection: ConnectionId,
+        record: Option<Record>,
+    ) {
+        debug!(node = self.node.name, "Kad Received PutRecord");
+        if record.is_none() {
+            warn!("Received PutRecord with no record");
+            return;
+        }
+        let record = record.unwrap();
+        let id = record.key.clone();
+
+        // save record to filem should use a VAULT here instead
+        self.put_record_into_vault(record);
+
+        // Start a query to be providing the file ID in kademlia
+        let qid = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .start_providing(id.clone());
+        if qid.is_err() {
+            debug!(
+                node = self.node.name,
+                err = format!("{qid:?}"),
+                "Failed to start providing file"
+            );
+            return;
+        }
+        let qid = qid.unwrap();
+
+        // We can't await for a response here because it would block the event loop
+        let (response_sender, _) = oneshot::channel();
+        self.behaviour
+            .pending_start_providing
+            .insert(qid, response_sender);
+    }
+
+    fn handle_inbound_request_add_provider(&mut self, record: Option<ProviderRecord>) {
+        match record {
+            Some(record) => {
+                self.swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .store_mut()
+                    .add_provider(record.clone())
+                    .ok(); // TODO What if the providers amount is exceeded? How to ensure only the closest one are kept?
+                info!(
+                    node = self.node.name,
+                    provider = record.provider.to_base58(),
+                    record = bs58::encode(&record.key).into_string(),
+                    "Received AddProvider"
+                );
+                self.print_providers(&record.key);
             }
-            _ => {}
+            None => {
+                warn!(node = self.node.name, "Received AddProvider with no record");
+            }
+        }
+    }
+
+    fn handle_inbound_request_get_provider(
+        &mut self,
+        _num_closer_peers: usize,
+        _num_provider_peers: usize,
+    ) {
+        debug!(node = self.node.name, "Kad Received GetProvider")
```

Comment:
Don't we want to handle that request?

Reply:
There is no need to do anything more here other than logging. This single trace might be droppped too, but I figured it would be helpful for me in the debugging process.